tests/src/ExistingSiteJavascript/CollaboraIntegrationTest.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Tests\collabora_online\ExistingSiteJavascript;
+
+use Behat\Mink\Element\Element;
+use Behat\Mink\Element\NodeElement;
+use Drupal\file\Entity\File;
+use Drupal\media\Entity\Media;
+use Drupal\media\MediaInterface;
+use weitzman\DrupalTestTraits\ExistingSiteSelenium2DriverTestBase;
+
+/**
+ * Test for Collabora Online editors embedded in or accessed from Drupal.
+ *
+ * To make this a regular FunctionalJavascript test, the WOPI request from
+ * Collabora to Drupal would need to send the SIMPLETEST_USER_AGENT cookie that
+ * tells Drupal that the request should be handled by the test installation,
+ * rather than the regular/existing installation.
+ *
+ * @see \drupal_valid_test_ua()
+ */
+class CollaboraIntegrationTest extends ExistingSiteSelenium2DriverTestBase {
+
+    /**
+     * Tests the Collabora editor in readonly mode.
+     */
+    public function testCollaboraPreview(): void {
+        // @todo Make sure there is always a document media type.
+        $user = $this->createUser([
+            'preview document in collabora',
+        ]);
+        $this->drupalLogin($user);
+        $text_file_basename = 'shopping-list';
+        $text_file_content = 'Chocolate, pickles';
+        $media = $this->createDocumentMedia('Shopping list', $text_file_basename, $text_file_content);
+        $this->drupalGet('/cool/view/' . $media->id());
+        $assert_session = $this->assertSession();
+        $this->getSession()->switchToIFrame('collabora-online-viewer');
+        // Wait until different parts of the editor are loaded and have the
+        // expected text or values.
+        $this->waitForRequiredElement('css', 'canvas#document-canvas', 2000);
+        // Make sure the correct document was opened.
+        // Check the document name at the top of the editor.
+        $this->assertSame(
+            $text_file_basename . '.txt',
+            // Even when the element exists, it might not have the correct value
+            // yet.
+            $this->waitForRequiredElement('css', 'input#document-name-input')->waitFor(
+                1000,
+                fn (Element $element) => $element->getValue(),
+            ),
+        );
```

Comment:
Avoid nesting, it makes understanding which part failed and debugging more difficult.
```
$input = $assert_session-> waitForRequiredElement('css', 'input#document-name-input');
$value = $this->waitForElementValue($text_file_basename . '.txt', $input);
$this->assertNotNull($value, 'The value was not the expected filename.';